### PR TITLE
menu: allow the librarian to switch library

### PR DESCRIPTION
**Pull request by Garfield-fr**
Enables the librarian to change from one libary to another
if he or she works in several libraries.

* Fixes patron view to display multiple libraries.
* Fix populate of select in the routing system.
* closes rero/rero-ils#1510

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>
Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Dependencies

* https://github.com/rero/ng-core/pull/306

## How to test?

- Check library switch with a system_librarian (All library of current organisation)
- Check library switch with a librarian with multiple libraries (Show only libraries attached)
- Check if the library switch menu is present if the librarian has only one library.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message propostion:

```
menu: allow the librarian to switch library

Enables the librarian to change from one libary to another
if he or she works in several libraries.

* Fixes patron view to display multiple libraries.
```

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

File: projects/admin/src/app/service/patron.service.ts [hunk 1]
```diff
@@ -78,6 +78,7 @@ export class PatronService {
             }
             case 1: {
               const patron = new User(response.hits.hits[0].metadata);
+
```
Comment (lauren-d) [inline projects/admin/src/app/service/patron.service.ts:81]: cosmetic: could be removed ?

**Review by lauren-d — APPROVED** [review]

File: projects/admin/src/app/service/patron.service.ts (hunk 1 above)
Comment (Garfield-fr) [reply projects/admin/src/app/service/patron.service.ts]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        librariesKeys.map((key: any) => {
+          this._librariesMenu[key].removeSuffix();
+        });
+        this._setActive(this._librariesMenu[user.currentLibrary]);
+      }
+    });
+
+    // when a library resource is created
+    // --> if a new library is created, then regenerate the menu
+    this._recordService.onCreate$.subscribe((event: RecordEvent) => {
```
Comment (jma) [inline projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:187]: You need to unsubscribe in the ngOnDestroy. Same for other subscriptions.

**Review by zannkukai — APPROVED** [review]
Some cosmetic changes only

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    ).subscribe((libraries: any) => {
+      this._generateMenu(libraries);
+    });
+  }
+
+  /** Librarian libraries */
+  private _librarianLibraries() {
+    const librariesPid = this._user.libraries.map(library => {
+      return library.pid;
+    });
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:96]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    this._executeQuery(
+      this._libraryService.findByLibrariesPidAndOrderBy$(librariesPid, 'name')
+    ).subscribe((libraries: any) => {
+      this._generateMenu(libraries);
+    });
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:101]: done

File: projects/admin/src/app/menu/menu-switch-library/menu-switch-library.component.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  ) { }
+
+  /**
+   * Event on menu click
+   * @param event - MenuItem
+   */
+  eventMenuClick(event: MenuItem) {
+    // TODO: Implement this with a guard (implement in ng-core and ui)
+    const url = this._router.url;
+    const urlParams = url.split('/').filter(n => n);
```
Comment (jma) [inline projects/admin/src/app/menu/menu-switch-library/menu-switch-library.component.ts:68]: perhaps you can use: https://angular.io/api/router/UrlSerializer

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        // Generate menu
+        this._generate();
+      } else {
+        // Use proxy to update menu
+        this._currentLibrary.setName(
+          this._librariesMenu[user.currentLibrary].getExtra('code')
+        );
+        librariesKeys.map((key: any) => {
+          this._librariesMenu[key].removeSuffix();
+        });
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:180]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+      this._generateMenu(libraries);
+    });
+  }
+
+  /**
+   * Execute query with an Observable
+   * @param query - Observable
+   * @return Observable
+   */
+  private _executeQuery(query: any) {
```
Comment (jma) [inline projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:109]: This method can have a more explicit name, for example which kind of query.

**Review by jma — APPROVED** [review]

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    }
+  }
+
+  /** System librarian libraries */
+  private _systemLibrarianLibraries() {
+    this._executeQuery(
+      this._libraryService.findAllOrderBy$('name')
+    ).subscribe((libraries: any) => {
+      this._generateMenu(libraries);
+    });
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:89]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch.service.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+   * @param libraryPid: the library pid that user would used
+   */
+  switch(libraryPid: string): void {
+    const user = this._userService.user;
+    // If the person is a librarian, we check the existence
+    // of the library in her libraries
+    if (!user.hasRole('system_librarian')) {
+      const libraries = user.libraries.map(library => {
+        return String(library.pid);
+      });
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch.service.ts:64]: done

**Review by lauren-d — APPROVED** [review]

File: projects/admin/src/app/record/detail-view/patron-detail-view/patron-detail-view.component.html [hunk 10]
```diff
@@ -82,30 +87,43 @@ <h1 class="mb-3">{{ patron.last_name }} {{ patron.first_name }}</h1>
...
-        <dt class="col-sm-3 offset-sm-2 offset-md-0">
-          <ng-container *ngIf="patron.roles.length === 1; else roles" translate>Role</ng-container>
-          <ng-template #roles translate>Roles</ng-template>:
+        <dt class="col-sm-3 offset-sm-2 offset-md-0 label-title" translate>
+          <ng-container *ngIf="patron.roles.length === 1; else roles" translate>
+            Role
+          </ng-container>
+          <ng-template #roles translate>
+            Roles
+          </ng-template>
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/record/detail-view/patron-detail-view/patron-detail-view.component.html:126]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+  /** menu is visible */
+  get visible() {
+    return this._user.hasRole('system_librarian')
+      ? true
+      : this._user.libraries.length > 1;
+  }
+
+  /** Get menu */
+  get menu() {
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:51]:
```
  get visible() {
    return this._user.hasRole('system_librarian')
      ? true
      : this._user.libraries.length > 1;
  }
```

File: projects/admin/src/app/menu/menu-switch-library/menu-switch-library.component.ts (hunk 5 above)
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/menu-switch-library.component.ts:68]: done

File: projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts (hunk 7 above)
Comment (Garfield-fr) [reply projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts:109]: done

**Pull request merged**